python/pyarrow/_s3.pyx

```diff
@@ -0,0 +1,100 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# cython: language_level = 3
+
+import six
+
+from pyarrow.compat import frombytes, tobytes
+from pyarrow.includes.common cimport *
+from pyarrow.includes.libarrow cimport *
+from pyarrow.includes.libarrow_s3 cimport *
+from pyarrow._fs cimport FileSystem
+from pyarrow.lib cimport check_status
+
+
+cpdef enum S3LogLevel:
+    Off = <int8_t> CS3LogLevel_Off
+    Fatal = <int8_t> CS3LogLevel_Fatal
+    Error = <int8_t> CS3LogLevel_Error
+    Warn = <int8_t> CS3LogLevel_Warn
+    Info = <int8_t> CS3LogLevel_Info
+    Debug = <int8_t> CS3LogLevel_Debug
+    Trace = <int8_t> CS3LogLevel_Trace
+
+
+def initialize_s3(S3LogLevel log_level=S3LogLevel.Error):
+    cdef CS3GlobalOptions options
+    options.log_level = <CS3LogLevel> log_level
+    check_status(CInitializeS3(options))
+
+
+def finalize_s3():
+    check_status(CFinalizeS3())
+
+
+cdef class S3FileSystem(FileSystem):
+    """S3-backed FileSystem implementation
+
+    Note: S3 buckets are special and the operations available on them may be
+          limited or more expensive than desired.
+
+    Parameters
+    ----------
+    access_key: str, default None
+        AWS Access Key ID. Pass None to use the standard AWS environment
+        variables and/or configuration file.
+    secret_key: str, default None
+        AWS Secret Access key. Pass None to use the standard AWS environment
+        variables and/or configuration file.
+    region: str, default 'us-east-1'
+        AWS region to connect to.
+    scheme: str, default 'https'
+        S3 connection transport scheme.
+    endpoint_override: str, default None
+        Override region with a connect string such as "localhost:9000"
+    background_writes: boolean, default True
+        Whether OutputStream writes will be issued in the background, without
+        blocking.
+    """
+
+    cdef:
+        CS3FileSystem* s3fs
+
+    def __init__(self, access_key=None, secret_key=None, region='us-east-1',
```

Comment:
I'd still rather have a separate `S3Options` Python class as in C++, especially since there are different ways to configure authentication.

Reply:
IMO keyword arguments are more pythonic, and we use keyword arguments throughout the yarrow codebase instead of `*Options` objects. How about exposing different configuration ways as classmethod factories?

Reply:
Well, we have `*Options` classes in `pyarrow.csv` already. And an options class is easier to pass around (e.g. in the datasets layer).

Reply:
ping @pitrou